### PR TITLE
draft of CleanCartTest.cs

### PR DESCRIPTION
**Pull request by VolhaKuzmiankova**

### CODE REVIEW

**Comment by VolhaKuzmiankova** [comment]
[Create test for delete test case](https://trello.com/c/Lvc5GAjZ/20-create-test-for-delete-test-case)

File: OnlinerByTestFramework/Constants/Asserts.cs [hunk 1]
```diff
@@ -0,0 +1,7 @@
+﻿namespace OnlinerByTestFramework.Constants
+{
+    public static class Asserts
```
Comment (mashadolya) [inline OnlinerByTestFramework/Constants/Asserts.cs]: Messages

File: OnlinerByTestFramework/Constants/DataForLocator.cs [hunk 2]
```diff
@@ -0,0 +1,8 @@
+﻿namespace OnlinerByTestFramework.Constants
+{
+    public static class DataForLocator
+    {
+        public const string TypeOfGoods = "Телевизоры";
+        public const string TvModel = "ЖК-телевизор Samsung QE55Q70AAU";
+    }
```
Comment (mashadolya) [inline OnlinerByTestFramework/Constants/DataForLocator.cs]: naming

File: OnlinerByTestFramework/Pages/CartPage.cs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+        }
+
+        public string Message()
+        {
+           return Waits.GetElementVisible(MessageSelector).Text;
+            
+        }
+        
+    }
+}
```
Comment (mashadolya) [inline OnlinerByTestFramework/Pages/CartPage.cs]: formatting

File: OnlinerByTestFramework/Pages/CartPage.cs [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+        }
+
+        public CartPage DeleteGood()
+        {
+            _action.MoveToElementAndClick(DeleteButtonSelector);
+
+            return this;
+        }
+
+        public string Message()
```
Comment (mashadolya) [inline OnlinerByTestFramework/Pages/CartPage.cs]: Get

File: OnlinerByTestFramework/Pages/CatalogSamsungPage.cs [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+{
+    public class CatalogSamsungPage : BasePage
+    {
+
+        public CatalogSamsungPage(IWebDriver driver) : base(driver, GetLocator(DataForLocator.TvModel), PageName.CatalogPage)
+        {
+        }
+        private static By GetLocator(string model)
+        {
+            return By.XPath($"//*[contains (text(), '{model}')]");
```
Comment (mashadolya) [inline OnlinerByTestFramework/Pages/CatalogSamsungPage.cs]: provide more specific locator

Comment (mashadolya) [reply OnlinerByTestFramework/Pages/CatalogSamsungPage.cs]: not sure that we need this method

File: OnlinerByTestFramework/Pages/GoodsPage.cs [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+        private static readonly By BuyButtonSelector = By.LinkText("В корзину");
+
+        private static readonly By ToCartButtonSelector = By.LinkText("В корзине");
+
+
+        public GoodsPage(IWebDriver driver) : base(driver, BuyButtonSelector, PageName.GoodsPage)
+        {
+        }
+
+        public GoodsPage ClickButton()
```
Comment (mashadolya) [inline OnlinerByTestFramework/Pages/GoodsPage.cs]: ClickBuyButton

File: OnlinerByTestFramework/Steps/ClenCartSteps.cs [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        
+        public CleanCartSteps(IWebDriver driver) : base(driver)
+        {
+            _homePage = new HomePage(driver);
+            _catalogSamsung = new CatalogSamsungPage(driver);
+            _goodsPage = new GoodsPage(driver);
+            _cartPage = new CartPage(driver);
+        }
+
+        public CleanCartSteps OpenCatalog()
```
Comment (mashadolya) [inline OnlinerByTestFramework/Steps/ClenCartSteps.cs]: path TypeOfGoods as parameter

File: OnlinerByTestFramework/Steps/ClenCartSteps.cs [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+        {
+            _homePage.OpenCatalog(Constants.DataForLocator.TypeOfGoods);
+            
+            return this;
+        }
+
+        public CleanCartSteps ChooseModel()
```
Comment (mashadolya) [inline OnlinerByTestFramework/Steps/ClenCartSteps.cs]: see above

File: OnlinerByTestFramework/Steps/ClenCartSteps.cs [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+        public CleanCartSteps DeleteSelectedGood()
+        {
+            _cartPage.DeleteGood();
+            
+            return this;
+        }
+
+        public CleanCartSteps MessageVerification()
+        {
+            Assert.StartsWith(Constants.Asserts.DeleteMessage, _cartPage.Message());
```
Comment (mashadolya) [inline OnlinerByTestFramework/Steps/ClenCartSteps.cs]: atual as parameter 
`MessageShouldBe(expected)`

File: OnlinerByTestFramework/Tests/CleanCartTest.cs [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+
+
+        public CleanCartTest(TestFixture fixture)
+        {
+            _fixture = fixture;
+            _cartPage = new CartPage(_fixture.Driver);
+        }
+
+        [Fact]
+        public void DeleteSelectedGood_FromCart()
```
Comment (mashadolya) [inline OnlinerByTestFramework/Tests/CleanCartTest.cs]: naming

File: OnlinerByTestFramework/Tests/CleanCartTest.cs [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+        {
+            var cleanCartSteps = new CleanCartSteps(_fixture.Driver);
+
+            cleanCartSteps.OpenCatalog()
```
Comment (mashadolya) [inline OnlinerByTestFramework/Tests/CleanCartTest.cs]: Move logic with opening goods to the `NavigationBar.cs` 
NavigationBar should be located on `CatalogPage.cs`

File: OnlinerByTestFramework/Utils/Waits.cs [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+using OpenQA.Selenium.Support.UI;
+using SeleniumExtras.WaitHelpers;
+
+namespace OnlinerByTestFramework.Utils
+{
+    public class Waits
+    {
+        private WebDriverWait _wait;
+
+        public Waits(IWebDriver driver)
```
Comment (mashadolya) [inline OnlinerByTestFramework/Utils/Waits.cs]: static

File: OnlinerByTestFramework/Utils/Action.cs [hunk 13]
```diff
@@ -0,0 +1,24 @@
+﻿using OpenQA.Selenium;
+using OpenQA.Selenium.Interactions;
+
+namespace OnlinerByTestFramework.Utils
+{
+    public class Action
```
Comment (mashadolya) [inline OnlinerByTestFramework/Utils/Action.cs]: static

File: OnlinerByTestFramework/Utils/Waits.cs [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+            }
+            catch (NoSuchElementException)
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        public IWebElement GetClickable(By by)
```
Comment (mashadolya) [inline OnlinerByTestFramework/Utils/Waits.cs]: `WaitToBeClickable`

**Review by mashadolya — CHANGES_REQUESTED** [review]

File: OnlinerByTestFramework/Utils/Waits.cs [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+
+namespace OnlinerByTestFramework.Utils
+{
+    public class Waits
+    {
+        private WebDriverWait _wait;
+
+        public Waits(IWebDriver driver)
+        {
+            _wait = new WebDriverWait(driver, TimeSpan.FromSeconds(3));
```
Comment (mashadolya) [inline OnlinerByTestFramework/Utils/Waits.cs]: Move timeout to config file

**Review by mashadolya — CHANGES_REQUESTED** [review]

File: .idea/.idea.OnlinerByTestFramework/.idea/encodings.xml [hunk 16]
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project version="4">
+  <component name="Encoding" addBOMForNewFiles="with BOM under Windows, with no BOM otherwise" />
+</project>
```
Comment (mashadolya) [inline .idea/.idea.OnlinerByTestFramework/.idea/encodings.xml:4]: gitignore

File: OnlinerByTestFramework/Pages/Base/BasePage.cs [hunk 17]
```diff
@@ -1,17 +1,29 @@
...
+        protected readonly IWebDriver Driver;
+        private readonly By locator;
+        private readonly string _pageName;
 
-        protected IWebDriver _driver;
+        protected BasePage(IWebDriver driver, By locator, string pageName)
+        {
+            Driver = driver;
+            _pageName = pageName;
+            this.locator = locator;
```
Comment (mashadolya) [inline OnlinerByTestFramework/Pages/Base/BasePage.cs]: delete this

File: OnlinerByTestFramework/Pages/HomePage.cs [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+    {
+        private static readonly By CatalogSelector = By.XPath($"//span[contains(text(),'{TypeOfGoods.ProductTv}')]");
+
+        public HomePage(IWebDriver driver) : base(driver, CatalogSelector, PageName.HomePage)
+        {
+        }
+
+        public CatalogSamsungPage OpenCatalog()
+        {
+            Driver.FindElement(CatalogSelector).Click();
```
Comment (mashadolya) [inline OnlinerByTestFramework/Pages/HomePage.cs]: create wrapper

File: OnlinerByTestFramework/Pages/LoginPage.cs [hunk 19]
```diff
@@ -7,37 +8,51 @@ public class LoginPage : BasePage
...
 
-        private static readonly By PasswordSelector = By.XPath("//*[@placeholder ='Пароль']");
+        private static readonly By PasswordSelector = By.XPath("//input[@placeholder ='Пароль']");
 
-        private static readonly By LoginButtonSelector = By.XPath("//*[@class ='auth-form__control auth-form__control_condensed-additional']");
+        private static readonly By LoginButtonSelector =
+            By.XPath("//*[@class ='auth-form__control auth-form__control_condensed-additional']");
 
-        public LoginPage(IWebDriver driver) : base(driver)
+        private static readonly By GoodsTypeSelector = By.XPath("//span[contains(text(),'Телевизоры')]");
```
Comment (mashadolya) [inline OnlinerByTestFramework/Pages/LoginPage.cs]: Телевизоры to const

File: OnlinerByTestFramework/Steps/ClenCartSteps.cs [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+        }
+
+        public CleanCartSteps OpenCatalog(string product)
+        {
+            _homePage.OpenCatalog();
+
+            return this;
+        }
+
+        public CleanCartSteps ChooseModel(string model)
```
Comment (mashadolya) [inline OnlinerByTestFramework/Steps/ClenCartSteps.cs]: ?model

File: OnlinerByTestFramework/Steps/ClenCartSteps.cs [hunk 21]
```diff
@@ -0,0 +1,59 @@
...
+        {
+            _catalogSamsung.OpenGoodsPage();
+
+            return this;
+        }
+
+        public CleanCartSteps AddTheGood()
```
Comment (mashadolya) [inline OnlinerByTestFramework/Steps/ClenCartSteps.cs]: AddTheGoodToCart

File: OnlinerByTestFramework/Tests/CleanCartTest.cs [hunk 22]
```diff
@@ -0,0 +1,30 @@
...
+using OnlinerByTestFramework.Steps;
+using Xunit;
+
+namespace OnlinerByTestFramework.Tests
+{
+    public class CleanCartTest : IClassFixture<TestFixture>
+    {
+        private readonly TestFixture _fixture;
+
+
```
Comment (mashadolya) [inline OnlinerByTestFramework/Tests/CleanCartTest.cs]: delete extra line

File: OnlinerByTestFramework/Tests/CleanCartTest.cs [hunk 23]
```diff
@@ -0,0 +1,30 @@
...
+        public CleanCartTest(TestFixture fixture)
+        {
+            _fixture = fixture;
+        }
+
+        [Fact]
```
Comment (mashadolya) [inline OnlinerByTestFramework/Tests/CleanCartTest.cs]: add description

File: OnlinerByTestFramework/Tests/CleanCartTest.cs [hunk 24]
```diff
@@ -0,0 +1,30 @@
...
+            var cleanCartSteps = new CleanCartSteps(_fixture.Driver);
+
+            cleanCartSteps.OpenCatalog(TypeOfGoods.ProductTv)
+                .ChooseModel(TypeOfGoods.TvModel)
+                .AddTheGood()
+                .DeleteSelectedGood()
+                .MessageShouldBe(Messages.DeleteMessage);
+        }
+    }
+}
```
Comment (mashadolya) [inline OnlinerByTestFramework/Tests/CleanCartTest.cs]: configure IDE to ad blank line

File: OnlinerByTestFramework/Utils/Waits.cs [hunk 25]
```diff
@@ -0,0 +1,40 @@
+﻿using System;
+using OpenQA.Selenium;
+using OpenQA.Selenium.Support.UI;
+using SeleniumExtras.WaitHelpers;
+
+namespace OnlinerByTestFramework.Utils
+{
+    public static class Waits
+    {
+        public static bool IsElementVisible(IWebDriver driver, By by)
```
Comment (mashadolya) [inline OnlinerByTestFramework/Utils/Waits.cs]: Startup.AppSettings.Configuration.WaitForTimeout as default parameter

**Review by mashadolya — CHANGES_REQUESTED** [review]

File: OnlinerByTestFramework/Pages/CatalogPage.cs [hunk 26]
```diff
@@ -0,0 +1,24 @@
+﻿using OnlinerByTestFramework.Constants;
+using OnlinerByTestFramework.Pages.Base;
+using OnlinerByTestFramework.Wrappers;
+using OpenQA.Selenium;
+
+namespace OnlinerByTestFramework.Pages
+{
+    public class CatalogPage : BasePage
```
Comment (Maria-Dolia) [inline OnlinerByTestFramework/Pages/CatalogPage.cs]: Shold be common goods page

File: OnlinerByTestFramework/Pages/HomePage.cs [hunk 27]
```diff
@@ -1,22 +1,24 @@
...
 using OnlinerByTestFramework.Pages.Base;
+using OnlinerByTestFramework.Wrappers;
 using OpenQA.Selenium;
 
 namespace OnlinerByTestFramework.Pages
 {
     public class HomePage : BasePage
     {
-        private static readonly By CatalogSelector = By.XPath($"//span[contains(text(),'{TypeOfGoods.ProductTv}')]");
+        private static readonly By CatalogSelector =
```
Comment (Maria-Dolia) [inline OnlinerByTestFramework/Pages/HomePage.cs]: as below

File: OnlinerByTestFramework/Pages/LoginPage.cs [hunk 28]
```diff
@@ -15,43 +16,35 @@ public class LoginPage : BasePage
...
-        private static readonly By GoodsTypeSelector = By.XPath("//span[contains(text(),'Телевизоры')]");
 
         public LoginPage(IWebDriver driver) : base(driver, ButtonSelector, PageName.LoginPage)
         {
         }
 
         public LoginPage OpenForm()
         {
-            Driver.FindElement(ButtonSelector).Click();
+            new WebElement("Button", Driver, ButtonSelector).Click();
```
Comment (Maria-Dolia) [inline OnlinerByTestFramework/Pages/LoginPage.cs]: provide specific name

File: OnlinerByTestFramework/Wrappers/WebElement.cs [hunk 29]
```diff
@@ -0,0 +1,62 @@
...
+
+namespace OnlinerByTestFramework.Wrappers
+{
+    public class WebElement
+    {
+        private IWebDriver _driver;
+        private By _by;
+        private IWebElement _element;
+
+
```
Comment (Maria-Dolia) [inline OnlinerByTestFramework/Wrappers/WebElement.cs]: ????

File: OnlinerByTestFramework/Wrappers/WebElement.cs [hunk 30]
```diff
@@ -0,0 +1,62 @@
...
+using OnlinerByTestFramework.Exceptions;
+using OpenQA.Selenium;
+
+namespace OnlinerByTestFramework.Wrappers
+{
+    public class WebElement
+    {
+        private IWebDriver _driver;
+        private By _by;
+        private IWebElement _element;
```
Comment (Maria-Dolia) [inline OnlinerByTestFramework/Wrappers/WebElement.cs]: readonly

File: OnlinerByTestFramework/Wrappers/WebElement.cs [hunk 31]
```diff
@@ -0,0 +1,62 @@
...
+                throw;
+            }
+        }
+
+        public ReadOnlyCollection<IWebElement> FindElements(By by)
+        {
+            return _driver.FindElements(by);
+        }
+
+
```
Comment (Maria-Dolia) [inline OnlinerByTestFramework/Wrappers/WebElement.cs]: ???

File: OnlinerByTestFramework/Wrappers/WebElement.cs [hunk 32]
```diff
@@ -0,0 +1,62 @@
...
+        {
+            FindElement(_by).SendKeys(text);
+        }
+
+        public void Submit()
+        {
+            FindElement(_by).Submit();
+        }
+
+
```
Comment (Maria-Dolia) [inline OnlinerByTestFramework/Wrappers/WebElement.cs]: ?

**Review by Maria-Dolia — CHANGES_REQUESTED** [review]

File: OnlinerByTestFramework/Constants/GoodsAndModels.cs [hunk 33]
```diff
@@ -10,7 +10,7 @@ public static class Products
         public static class Models
 
         {
-            public const string TvModel = "Samsung UE50TU7570U";
+            public const string TvModel = "Samsung QE55Q80AAU";
```
Comment (Maria-Dolia) [inline OnlinerByTestFramework/Constants/GoodsAndModels.cs]: pls move to test

File: OnlinerByTestFramework/Pages/CartPage.cs [hunk 34]
```diff
@@ -8,8 +8,8 @@ namespace OnlinerByTestFramework.Pages
     public class CartPage : BasePage
     {
         private static readonly By DeleteButtonSelector = By.XPath("//div[@class='cart-form__control'][1]");
-
-        private static readonly By MessageSelector = By.XPath($" //div[contains(text(),'{TypeOfGoods.Models.TvModel}')]");
+        
+        private static readonly By MessageSelector = By.XPath($" //div[contains(text(),'{GoodsAndModels.Models.TvModel}')]");
```
Comment (Maria-Dolia) [inline OnlinerByTestFramework/Pages/CartPage.cs]: do not use hardcoded tv model

File: OnlinerByTestFramework/Pages/HomePage.cs [hunk 35]
```diff
@@ -1,24 +1,32 @@
...
 namespace OnlinerByTestFramework.Pages
 {
     public class HomePage : BasePage
     {
-        private static readonly By CatalogSelector =
-            By.XPath($"//span[contains(text(),'{TypeOfGoods.Products.ProductTv}')]");
+        private static readonly By SearchInputSelector = By.ClassName("fast-search__input");
 
-        public HomePage(IWebDriver driver) : base(driver, CatalogSelector, PageName.HomePage)
+        private static readonly By CategoryOfGoodsSelector = By.CssSelector("li[data-bind*='search__result_active']:first-child");
```
Comment (Maria-Dolia) [inline OnlinerByTestFramework/Pages/HomePage.cs]: GoodsCategorySelector would be better

File: OnlinerByTestFramework/Pages/LoginPage.cs [hunk 36]
```diff
@@ -13,31 +13,28 @@ public class LoginPage : BasePage
...
-            new WebElement("Button", Driver, ButtonSelector).Click();
+            new WebElement("EnterButton", Driver, ButtonSelector).Click();
 
             return this;
         }
 
         public LoginPage TypeUsername(string username)
         {
-            new WebElement("Username", Driver, EmailSelector).SendKeys(username);
+            new WebElement("UsernameInput", Driver, EmailSelector).SendKeys(username);
```
Comment (Maria-Dolia) [inline OnlinerByTestFramework/Pages/LoginPage.cs:31]: pls use common naming approach for elements

File: OnlinerByTestFramework/Pages/SelectItemPage.cs [hunk 37]
```diff
@@ -5,11 +5,10 @@
...
 namespace OnlinerByTestFramework.Pages
 {
-    public class CatalogPage : BasePage
+    public class SelectItemPage : BasePage
     {
-        private static readonly By ModelSelector = By.XPath($"//span[contains (text(), '{TypeOfGoods.Models.TvModel}')]");
-
-        public CatalogPage(IWebDriver driver) : base(driver, ModelSelector, PageName.CatalogPage)
+        private static readonly By ModelSelector = By.XPath($"//span[contains (text(), '{GoodsAndModels.Models.TvModel}')]");
+        public SelectItemPage(IWebDriver driver) : base(driver, ModelSelector, PageName.CatalogPage)
```
Comment (Maria-Dolia) [inline OnlinerByTestFramework/Pages/SelectItemPage.cs]: blank line 
hardcoded model

File: OnlinerByTestFramework/Steps/AddToCartSteps.cs [hunk 38]
```diff
@@ -0,0 +1,31 @@
+﻿using OnlinerByTestFramework.Pages;
+using OnlinerByTestFramework.Steps.Base;
+using OpenQA.Selenium;
+
+namespace OnlinerByTestFramework.Steps
+{
+    public class AddToCart : BaseSteps
```
Comment (Maria-Dolia) [inline OnlinerByTestFramework/Steps/AddToCartSteps.cs]: naming

File: OnlinerByTestFramework/Steps/AddToCartSteps.cs [hunk 39]
```diff
@@ -0,0 +1,31 @@
...
+    {
+        private readonly SelectItemPage _selectItemPagePage;
+        private readonly GoodsPage _goodsPage;
+
+        public AddToCart(IWebDriver driver) : base(driver)
+        {
+            _selectItemPagePage = new SelectItemPage(driver);
+            _goodsPage = new GoodsPage(driver);
```
Comment (Maria-Dolia) [inline OnlinerByTestFramework/Steps/AddToCartSteps.cs]: I would recommend to initialize page in the method where it is used

**Review by Maria-Dolia — CHANGES_REQUESTED** [review]

**Review by Maria-Dolia — APPROVED** [review]

**Pull request merged**